### ISSUE TITLE
[FEATURE] Support token substitution in flow definitions

### ISSUE THREAD

**Issue by s3-4v** [marklogic/marklogic-data-hub#4929]
**Is your feature request related to a problem? Please describe.**

I would like to specify variable source and target databases in mapping steps in flows when deploying to various environments using token substitution; however, currently, this feature is not yet supported.

**Describe the solution you'd like**

Please support token replacement in flows, so that we can achieve the following effect for clients/MarkLogic customers, for example:
 
```
"sourceDatabase" : "%%mlAppName%%-STAGING",
"targetDatabase" : "%%mlAppName%%-FINAL"
```

**Describe alternatives you've considered**

According to Rob, the present work-around is use a custom Gradle task to manipulate the contents of the documents after they've been loaded. 

**Additional context**

On Stack Overflow, Rob recommended that I create a ticket to request this feature; he said, "It's a worthwhile one to add."

Thank you.

**Comment by srinathgit** [comment]
This feature has been implemented. It will be part of Datahub 5.4.0

**Comment by dmcassel** [comment]
[The FlowRunner interface](https://github.com/marklogic/marklogic-data-hub/blob/v5.2.4/marklogic-data-hub/src/main/java/com/marklogic/hub/flow/FlowRunner.java) has a bunch of comments on the `runFlow` methods saying:

```
@deprecated Starting in 5.2.0, prefer runFlow(FlowInputs), which does not depend on Spring or project files on the filesystem
```

However, [the 5.2.6 version of DHF still calls one of the deprecated `runFlow` methods in the `RunFlowTask`](https://github.com/marklogic/marklogic-data-hub/blob/v5.2.6/ml-data-hub-plugin/src/main/groovy/com/marklogic/gradle/task/RunFlowTask.groovy#L144). That means that if we have tokens in the flow or step definitions, we get errors when running a flow from gradle. 

Could we get a 5.2.7 that calls `flowRunner.runFlow(flowInputs)`?

**Comment by rjrudin** [comment]
Pinging @srinathgit and @dmcassel  on this - I don't believe this is an issue in 5.4 because while RunFlowTask is using a deprecated method, those deprecated methods are all going through the same "runFlow" method that loads the flow config from ML. Which perhaps means that while those methods can remain deprecated, the comment should be altered to indicate why they're deprecated - which is that we don't want to support N variations of the same overloaded method, and FlowInputs allows us to easily add optional arguments in the future. 

@srinathgit  Can you check that out?

**Comment by srinathgit** [comment]
The properties with tokens ("sourceDatabase" : "%%mlAppName%%-STAGING") in steps, flows, step definitions etc are translated while deploying them from gradle (using hubDeploy, mlDeploy or hubDeployUserArtifacts command) as per the original comment. That's what is implemented in 5.4.0. The translation doesn't happen at the time of running the flow, so there is no change as far as running the flow is concerned.

**Comment by dmcassel** [comment]
@srinathgit if I understand you comment, I think running a flow from gradle continues to use the version of the flow and step from the file system; is that what you're saying? 

@rjrudin unless I'm misreading the code (always possible), the difference between the versions of `runFlow` that take `flowName`, etc as separate parameters versus the one that takes the `FlowInputs` is that the latter [retrieves the flow and step definitions from MarkLogic](https://github.com/marklogic/marklogic-data-hub/blob/81eec483eccfa23486fc67740c86a7530b9de94f/marklogic-data-hub/src/main/java/com/marklogic/hub/flow/impl/FlowRunnerImpl.java#L158), while the former reads them from the file system. It looks to me like everything is in place to have the gradle task get the definitions from MarkLogic (where the tokens have been replaced), but the actual switch hasn't been made. I'm asking whether [this line](https://github.com/marklogic/marklogic-data-hub/blob/v5.2.6/ml-data-hub-plugin/src/main/groovy/com/marklogic/gradle/task/RunFlowTask.groovy#L144): 

    RunFlowResponse runFlowResponse = flowRunner.runFlow(flowName, steps, jobId, flowInputs.getOptions(), flowInputs.getStepConfig())

should change to this:

    RunFlowResponse runFlowResponse = flowRunner.runFlow(flowInputs)

`flowInputs` has already been built.

**Comment by srinathgit** [comment]
@dmcassel 

No, In 5.4.0, running a flow from gradle always fetches the flow, step from MarkLogic . Both the runFlow() methods use"flowManager.getFullFlow()" which fetches the flow from MarkLogic db.

https://github.com/marklogic/marklogic-data-hub/blob/develop/marklogic-data-hub/src/main/java/com/marklogic/hub/flow/impl/FlowRunnerImpl.java#L155-L179

**Comment by rjrudin** [comment]
I did a quick test to confirm that this is the case. I believe the deprecation comment is still accurate enough, as in 5.2, the deprecated runFlow methods do depend on the local project files. 

Closing this as we know it's being addressed in 5.4.0.
